src/coredump/coredump.c

```diff
@@ -560,6 +560,90 @@ static int compose_open_fds(pid_t pid, char **open_fds) {
         return 0;
 }
 
+static int get_process_ns(pid_t pid, const char *namespace, ino_t *ns) {
+        const char *p;
+        struct stat stbuf;
+        _cleanup_close_ int proc_ns_dir_fd;
+
+        p = procfs_file_alloca(pid, "ns");
+
+        proc_ns_dir_fd = open(p, O_DIRECTORY);
+        if (proc_ns_dir_fd < 0)
+                return -errno;
+
+        if (fstatat(proc_ns_dir_fd, namespace, &stbuf, /* flags */0) != 0)
+                return -errno;
+
+        *ns = stbuf.st_ino;
+        return 0;
+}
+
+/* Find the first process in the parent hierarchy which has different mount
+ * namespace and call it the pid's container parent. */
+static int get_process_container_parent_pid(pid_t pid, pid_t *container_pid) {
+        pid_t cpid = pid, ppid = 0;
+        ino_t proc_mntns;
+        int r = 0;
+
+        r = get_process_ns(pid, "mnt", &proc_mntns);
+        if (r != 0)
+                return r;
+
+        while (1) {
+                ino_t parent_mntns;
+
+                r = get_process_ppid(cpid, &ppid);
+                if (r != 0)
+                        return r;
+
+                r = get_process_ns(ppid, "mnt", &parent_mntns);
+                if (r != 0)
+                        return r;
+
+                if (proc_mntns != parent_mntns)
+                        break;
+
+                if (ppid == 1)
+                        return -ENOENT;
+
+                cpid = ppid;
+        }
+
+        *container_pid = ppid;
+        return 0;
+}
+
+/* Returns 0 on success.
+ * Returns a negative number on errors.
+ * Returns a positive number if there is not a process we can call the pid's
+ * container parent (the pid's process isn't 'containerized').
+ */
+static int get_process_container_parent_cmdline(pid_t pid, char** cmdline) {
+        int r = 0;
+        pid_t container_pid;
+        const char *proc_root_path;
+        struct stat root_stat, proc_root_stat;
+
+        /* To compare inodes of / and /proc/[pid]/root */
+        if (stat("/", &root_stat) < 0)
+                return -errno;
+
+        proc_root_path = procfs_file_alloca(pid, "root");
+        if (stat(proc_root_path, &proc_root_stat) < 0)
+                return -errno;
+
+        /* The process uses system root. */
+        if (proc_root_stat.st_ino == root_stat.st_ino)
+                return 1;
+
+        r = get_process_container_parent_pid(pid, &container_pid);
+        if (r != 0)
+                return r;
+
+        r = get_process_cmdline(container_pid, 0, false, cmdline);
+        return r;
+}
```

Comment:
hmm, so this function returns something in *cmdline in some cases of success, but not in others? I really would prefer if we'd initialize the return parameters in all success cases, and in this particular case to NULL if we have no string to return.

It's just a matter of following the same coding style everywhere: don't clobber return parameters on failure, and always initialize them on success. 

I think it would also make sense to return "0" when the return parameter was not initialized and "1" if it was (think: this is kinda a boolean that tells if it found something). The code currently does it the other way round...